### PR TITLE
Refactor array util

### PR DESCRIPTION
**Pull request by jovilius**
Hi @dlwh,

an attempt to remove duplications in `ArrayUtil` without affecting performances. 

Does it make sense?

### CODE REVIEW

File: math/src/main/scala/breeze/util/ArrayUtil.scala [hunk 1]
```diff
@@ -105,46 +97,33 @@ object ArrayUtil {
...
+        case x: Array[Double] => b.asInstanceOf[Array[Double]] -> true
+        case x: Array[Float] => b.asInstanceOf[Array[Float]] -> true
+        case _ => b -> false
+      }
+      if (floating) {
+        var ai = aoffset
+        var bi = boffset
+        var i = 0
+        while (i < alength) {
+          if (a(ai) != y(bi)) return false
```
Comment (dlwh) [inline math/src/main/scala/breeze/util/ArrayUtil.scala]: this boxes, unfortunately

File: math/src/main/scala/breeze/util/ArrayUtil.scala [hunk 2]
```diff
@@ -159,159 +138,56 @@ object ArrayUtil {
...
+        case x: Array[Byte] => b.asInstanceOf[Array[Byte]]
+        case x: Array[Boolean] => b.asInstanceOf[Array[Boolean]]
+        case x: Array[_] => b.asInstanceOf[Array[AnyRef]]
         case _ => throw new RuntimeException("shouldn't be here!")
       }
+      var ai = aoffset
+      var bi = boffset
+      var i = 0
+      while (i < alength) {
+        if (a(ai) != y(bi)) return false
```
Comment (dlwh) [inline math/src/main/scala/breeze/util/ArrayUtil.scala]: boxing :(

**Comment by dlwh** [comment]
unfortunately this causes boxing, and possibly even reflection. If you want to do something like this, I'd recommend using the `@expand` macro like in zeroSkippingHashCodeImpl

**Comment by jovilius** [comment]
Thanks, I guessed so!

**Comment by jovilius** [comment]
Thanks for the feedback.

**Pull request closed without merging**